File: internal/game/likes/catalog/catalog.go
// Package catalog contains the fixed rules and descriptions for the likes game.
package catalog

import (
	"bytes"
	"crypto/sha256"
	"embed"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"slices"
)

const DesignVersion = "0.17.0"
const SchemaVersion = 15
const RulesVersion = 1

//go:embed quick.json standard.json
var presets embed.FS

var ErrCatalog = errors.New("likes: invalid catalog")

type Cost struct {
	Energy  int64  `json:"energy"`
	Token   int64  `json:"token"`
	Payment string `json:"payment"`
	Gold    int64  `json:"gold,omitempty"`
}

type Effect struct {
	Kind          string `json:"kind"`
	Likes         int64  `json:"likes"`
	P             int64  `json:"p"`
	Q             int64  `json:"q"`
	N             int64  `json:"n"`
	BuffID        string `json:"buffId,omitempty"`
	ExtraBuffID   string `json:"extraBuffId,omitempty"`
	CacheTarget   string `json:"cacheTarget,omitempty"`
	Meme          string `json:"meme,omitempty"`
	Combo         int64  `json:"combo,omitempty"`
	RandomTargets bool   `json:"randomTargets,omitempty"`
	AuditTarget   string `json:"auditTarget,omitempty"`
}

type Effects struct {
	Base Effect `json:"base"`
	I    Effect `json:"I"`
	II   Effect `json:"II"`
}

type Skill struct {
	ID            string           `json:"id"`
	Owner         string           `json:"owner"`
	Kind          string           `json:"kind"`
	Name          string           `json:"name"`
	Energy        int64            `json:"energy"`
	Token         int64            `json:"token"`
	Payment       string           `json:"payment"`
	Image         int64            `json:"image"`
	MaxUses       *int64           `json:"maxUses"`
	Learn         int64            `json:"learn"`
	Copyable      bool             `json:"copyable"`
	Stable        bool             `json:"stable"`
	Meme          string           `json:"meme"`
	Note          string           `json:"note"`
	Effects       Effects          `json:"effects"`
	ResourceCosts map[string]int64 `json:"resourceCosts"`
	Gold          int64            `json:"gold,omitempty"`
}

func (skill Skill) Cost() Cost {
	return Cost{Energy: skill.Energy, Token: skill.Token, Payment: skill.Payment, Gold: skill.Gold}
}

type Buff struct {
	ID             string `json:"id"`
	Name           string `json:"name"`
	Kind           string `json:"kind"`
	Category       string `json:"category,omitempty"`
	Source         string `json:"source,omitempty"`
	P              int64  `json:"p"`
	Q              int64  `json:"q"`
	N              int64  `json:"n"`
	Cap            int64  `json:"cap"`
	Trigger        string `json:"trigger"`
	Expiry         string `json:"expiry"`
	Overwrite      string `json:"overwrite"`
	Target         string `json:"target"`
	Description    string `json:"description,omitempty"`
	Meme           string `json:"meme,omitempty"`
	StackGroup     string `json:"stackGroup,omitempty"`
	Reapply        string `json:"reapply,omitempty"`
	CacheScope     string `json:"cacheScope,omitempty"`
	RefreshOnCombo bool   `json:"refreshOnCombo,omitempty"`
}

type Allocation struct {
	Initial int64 `json:"initial"`
	Cap     int64 `json:"cap"`
}
type Resource struct {
	ID           string `json:"id"`
	Name         string `json:"name"`
	Unit         string `json:"unit"`
	Description  string `json:"description"`
	Meme         string `json:"meme"`
	Subscription bool   `json:"subscription,omitempty"`
	Upgrade      int64  `json:"upgrade,omitempty"`
}
type Role struct {
	ID         string                `json:"id"`
	Name       string                `json:"name"`
	Focus      string                `json:"focus"`
	Difficulty string                `json:"difficulty"`
	Note       string                `json:"note"`
	Weakness   string                `json:"weakness"`
	Overrides  map[string]int64      `json:"overrides"`
	Resources  map[string]Allocation `json:"resources"`
}
type Loadout struct {
	ID     string   `json:"id"`
	Role   string   `json:"role"`
	Name   string   `json:"name"`
	Skills []string `json:"skills"`
	Note   string   `json:"note"`
}
type Passive struct {
	ID          string `json:"id"`
	Name        string `json:"name"`
	Kind        string `json:"kind"`
	P           int64  `json:"p"`
	Q           int64  `json:"q"`
	BuffID      string `json:"buffId,omitempty"`
	Description string `json:"description"`
	Meme        string `json:"meme"`
}
type Harness struct {
	ID          string   `json:"id"`
	Name        string   `json:"name"`
	ActiveSlots int      `json:"activeSlots"`
	Passives    []string `json:"passives"`
	Description string   `json:"description"`
	Meme        string   `json:"meme"`
}
type ParamMeta struct {
	ID   string `json:"id"`
	Name string `json:"name"`
	Unit string `json:"unit"`
	Note string `json:"note"`
}
type Rules struct {
	CacheWindow   string `json:"cacheWindow"`
	UniqueSamples bool   `json:"uniqueSamples"`
	StrictSamples bool   `json:"strictSamples"`
	ImageShortage string `json:"imageShortage"`
}
type Config struct {
	SchemaVersion int              `json:"schemaVersion"`
	Mode          string           `json:"mode"`
	Name          string           `json:"name"`
	Parameters    map[string]int64 `json:"parameters"`
	ParamMeta     []ParamMeta      `json:"paramMeta"`
	Rules         Rules            `json:"rules"`
	Roles         []Role           `json:"roles"`
	Skills        []Skill          `json:"skills"`
	Buffs         []Buff           `json:"buffs"`
	Loadouts      []Loadout        `json:"loadouts"`
	Resources     []Resource       `json:"resources"`
	Harnesses     []Harness        `json:"harnesses"`
	Passives      []Passive        `json:"passives"`
}

// Load returns a fresh snapshot; caller mutations never change another game.
func Load(mode string) (Config, string, error) {
	if mode != "quick" && mode != "standard" {
		return Config{}, "", ErrCatalog
	}
	body, err := presets.ReadFile(mode + ".json")
	if err != nil {
		return Config{}, "", err
	}
	var config Config
	decoder := json.NewDecoder(bytes.NewReader(body))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&config); err != nil {
		return Config{}, "", fmt.Errorf("%w: %v", ErrCatalog, err)
	}
	if config.Mode != mode {
		return Config{}, "", ErrCatalog
	}
	if err := config.Validate(); err != nil {
		return Config{}, "", err
	}
	// Pin the shared-energy exception, settlement split and manual casting rule.
	hash := sha256.New()
	hash.Write([]byte("likes@1;positive-energy-overload;separate-round-start;manual-main-unless-stunned;overload-state\n"))
	hash.Write(body)
	return config, hex.EncodeToString(hash.Sum(nil)), nil
}

var timedKinds = []string{"AMPLIFY", "SUPPRESS", "TOKEN_TAX", "NONBASIC_TAX", "SAVE_ENERGY", "API_DISCOUNT", "REGULATOR"}

// Validate checks the fixed content's references and the termination assumptions
// that keep a complete round and its event history within the service bounds.
func (c Config) Validate() error {
	if c.SchemaVersion != SchemaVersion || len(c.Roles) != 5 || len(c.Skills) != 48 || len(c.Buffs) != 46 || len(c.Resources) != 1 || len(c.Harnesses) != 8 || len(c.Passives) != 8 {
		return ErrCatalog
	}
	if c.Mode != "quick" && c.Mode != "standard" || c.Rules != (Rules{CacheWindow: "round", UniqueSamples: true, StrictSamples: true, ImageShortage: "illegal"}) {
		return ErrCatalog
	}
	for _, value := range c.Parameters {
		if value < 0 || value > 1_000_000 {
			return ErrCatalog
		}
	}
	if c.Parameters["PREP_MAX"] != 2 || c.Parameters["INSERT_CAP"] != 1 || c.Parameters["POWER_GENERATION"] != 10 || c.Parameters["TURN_SECONDS"] != 20 || c.Parameters["TOKEN_FLOOR"] != 1 || c.Parameters["ENERGY_FLOOR"] != 1 {
		return ErrCatalog
	}
	if c.Parameters["ENERGY_START"] > c.Parameters["ENERGY_CAP"] || c.Parameters["SUB_START"] != c.Parameters["BURST_CAP"]*2 || c.Parameters["SUB_TOTAL_UPGRADE"] != c.Parameters["SUB_BURST_UPGRADE"]*2 {
		return ErrCatalog
	}
	if c.Mode == "quick" && (c.Parameters["MAX_ROUNDS"] != 25 || c.Parameters["TARGET_LIKES"] != 60) || c.Mode == "standard" && (c.Parameters["MAX_ROUNDS"] != 75 || c.Parameters["TARGET_LIKES"] != 300) {
		return ErrCatalog
	}
	roles, buffs, skills, passives := map[string]Role{}, map[string]Buff{}, map[string]Skill{}, map[string]Passive{}
	for _, role := range c.Roles {
		if _, exists := roles[role.ID]; exists || !slices.Contains([]string{"ChatGPT", "Claude", "Gemini", "GLM", "DeepSeek"}, role.ID) {
			return ErrCatalog
		}
		roles[role.ID] = role
	}
	states := 0
	for _, buff := range c.Buffs {
		if _, exists := buffs[buff.ID]; exists || buff.ID == "" {
			return ErrCatalog
		}
		buffs[buff.ID] = buff
		for _, value := range []int64{buff.P, buff.Q, buff.N, buff.Cap} {
			if value < 0 || value > 1_000 {
				return ErrCatalog
			}
		}
		if buff.Category == "state" {
			states++
			if buff.Kind != "OVERLOAD" && buff.Kind != "SPEED_MODE" {
				return ErrCatalog
			}
		}
		if buff.Kind == "COMBO" && (buff.P != 3 || buff.Q != 1) {
			return ErrCatalog
		}
		if buff.Kind == "CACHE" && buff.Cap != 3 {
			return ErrCatalog
		}
		if buff.Reapply == "add" && slices.Contains(timedKinds, buff.Kind) && buff.Cap != 4 {
			return ErrCatalog
		}
	}
	if states != 2 || c.Resources[0].ID != "R_IMAGE" || !c.Resources[0].Subscription {
		return ErrCatalog
	}
	variants := 0
	for _, sk := range c.Skills {
		if _, exists := skills[sk.ID]; exists || sk.ID == "" {
			return ErrCatalog
		}
		skills[sk.ID] = sk
		if sk.Owner != "全局公共" {
			if _, exists := roles[sk.Owner]; !exists {
				return ErrCatalog
			}
		}
		if !slices.Contains([]string{"basic", "normal", "special", "ultimate"}, sk.Kind) || !slices.Contains([]string{"mix", "api", "sub"}, sk.Payment) {
			return ErrCatalog
		}
		for _, value := range []int64{sk.Energy, sk.Token, sk.Image, sk.Learn, sk.Gold} {
			if value < 0 || value > 1_000_000 {
				return ErrCatalog
			}
		}
		if sk.MaxUses != nil && (*sk.MaxUses < 0 || *sk.MaxUses > 100) {
			return ErrCatalog
		}
		for key, value := range sk.ResourceCosts {
			if key != "R_IMAGE" || value < 0 || value > 2 {
				return ErrCatalog
			}
		}
		variants++
		if sk.Copyable {
			variants += 2
		}
		for _, effect := range []Effect{sk.Effects.Base, sk.Effects.I, sk.Effects.II} {
			for _, value := range []int64{effect.Likes, effect.P, effect.Q, effect.N, effect.Combo} {
				if value < 0 || value > 1_000 {
					return ErrCatalog
				}
			}
			for _, key := range []string{effect.BuffID, effect.ExtraBuffID} {
				if key != "" {
					if _, exists := buffs[key]; !exists {
						return ErrCatalog
					}
				}
			}
			if effect.Kind == "INSERT" && effect.P != 1 {
				return ErrCatalog
			}
			progress := effect.Combo
			if effect.Kind == "COMBO" {
				progress += effect.P
			}
			if effect.ExtraBuffID != "" && effect.N > 0 && buffs[effect.ExtraBuffID].Kind == "COMBO" {
				progress++
			}
			if progress > 4 || effect.Kind == "CACHE_CONVERT" && progress != 0 {
				return ErrCatalog
			}
		}
	}
	if variants != 130 {
		return ErrCatalog
	}
	flash, distill := skills["GEM01"], skills["PUB41"]
	if flash.MaxUses != nil || flash.Effects.Base.Kind != "CACHE_COMBO" || flash.Effects.Base.N != 1 || flash.Effects.Base.ExtraBuffID == "" || buffs[flash.Effects.Base.ExtraBuffID].Kind != "COMBO" || flash.Effects.Base.Combo != 0 {
		return ErrCatalog
	}
	if distill.Copyable || distill.MaxUses != nil || distill.Payment != "api" || distill.Energy != 25 || distill.Token != 200 {
		return ErrCatalog
	}
	for _, sk := range c.Skills {
		for _, e := range []Effect{sk.Effects.Base, sk.Effects.I, sk.Effects.II} {
			if e.CacheTarget != "" {
				if _, exists := skills[e.CacheTarget]; !exists {
					return ErrCatalog
				}
			}
		}
	}
	for _, item := range c.Passives {
		if _, exists := passives[item.ID]; exists {
			return ErrCatalog
		}
		passives[item.ID] = item
		if item.BuffID != "" {
			if _, exists := buffs[item.BuffID]; !exists {
				return ErrCatalog
			}
		}
	}
	seen := map[string]bool{}
	for _, item := range c.Harnesses {
		if seen[item.ID] || item.ActiveSlots < 0 || item.ActiveSlots > 2 {
			return ErrCatalog
		}
		seen[item.ID] = true
		for _, key := range item.Passives {
			if _, exists := passives[key]; !exists {
				return ErrCatalog
			}
		}
	}
	return nil
}

File: web/src/user/games/likes/catalog.ts
import {
  booleanValue,
  enumValue,
  exactRecord,
  invalidResponse,
  safeInteger,
} from '../common/strict';
import { hashValue, list, nullable } from '../common/duel/normalize';
import { amount, dictionary, label, prose, unique } from './value';

export const ROLES = ['ChatGPT', 'Claude', 'Gemini', 'GLM', 'DeepSeek'] as const;
export type RoleID = (typeof ROLES)[number];
export const roleID = (v: unknown) => enumValue(v, ROLES, 'role');
export const SKILL_KINDS = ['basic', 'normal', 'special', 'ultimate'] as const;
export interface Effect {
  kind: string;
  likes: number;
  p: number;
  q: number;
  n: number;
  buffId?: string;
  extraBuffId?: string;
  cacheTarget?: string;
  meme?: string;
  combo?: number;
  randomTargets?: boolean;
  auditTarget?: string;
}
export interface Skill {
  id: string;
  owner: string;
  kind: (typeof SKILL_KINDS)[number];
  name: string;
  energy: number;
  token: number;
  payment: 'mix' | 'api' | 'sub';
  image: number;
  maxUses: number | null;
  learn: number;
  copyable: boolean;
  stable: boolean;
  meme: string;
  note: string;
  effects: { base: Effect; I: Effect; II: Effect };
  resourceCosts: Record<string, number>;
  gold?: number;
}
export interface Buff {
  id: string;
  name: string;
  kind: string;
  category?: string;
  source?: string;
  p: number;
  q: number;
  n: number;
  cap: number;
  trigger: string;
  expiry: string;
  overwrite: string;
  target: string;
  description?: string;
  meme?: string;
  stackGroup?: string;
  reapply?: string;
  cacheScope?: string;
  refreshOnCombo?: boolean;
}
export interface Role {
  id: RoleID;
  name: string;
  focus: string;
  difficulty: string;
  note: string;
  weakness: string;
  overrides: Record<string, number>;
  resources: Record<string, { initial: number; cap: number }>;
}
export interface Harness {
  id: string;
  name: string;
  activeSlots: number;
  passives: string[];
  description: string;
  meme: string;
}
export interface Passive {
  id: string;
  name: string;
  kind: string;
  p: number;
  q: number;
  buffId?: string;
  description: string;
  meme: string;
}
export interface Preset {
  id: string;
  role: string;
  name: string;
  skills: string[];
  note: string;
}
export interface Resource {
  id: string;
  name: string;
  unit: string;
  description: string;
  meme: string;
  subscription?: boolean;
  upgrade?: number;
}
export interface Parameter {
  id: string;
  name: string;
  unit: string;
  note: string;
}
export interface ModeCatalog {
  contentHash: string;
  mode: 'quick' | 'standard';
  name: string;
  parameters: Record<string, number>;
  paramMeta: Parameter[];
  roles: Role[];
  skills: Skill[];
  buffs: Buff[];
  harnesses: Harness[];
  passives: Passive[];
  loadouts: Preset[];
  resources: Resource[];
}
export interface LikesCatalog {
  contentHash: string;
  modes: { quick: ModeCatalog; standard: ModeCatalog };
}
const optionalText = (r: Record<string, unknown>, key: string) =>
  r[key] === undefined ? undefined : prose(r[key]);
function effect(value: unknown): Effect {
  const r = exactRecord(
    value,
    ['kind', 'likes', 'p', 'q', 'n'],
    ['buffId', 'extraBuffId', 'cacheTarget', 'meme', 'combo', 'randomTargets', 'auditTarget'],
  );
  return {
    kind: label(r.kind),
    likes: amount(r.likes),
    p: amount(r.p),
    q: amount(r.q),
    n: amount(r.n),
    buffId: optionalText(r, 'buffId'),
    extraBuffId: optionalText(r, 'extraBuffId'),
    cacheTarget: optionalText(r, 'cacheTarget'),
    meme: optionalText(r, 'meme'),
    combo: r.combo === undefined ? undefined : amount(r.combo),
    randomTargets:
      r.randomTargets === undefined ? undefined : booleanValue(r.randomTargets, 'random targets'),
    auditTarget: optionalText(r, 'auditTarget'),
  };
}
function skill(value: unknown): Skill {
  const r = exactRecord(
    value,
    [
      'id',
      'owner',
      'kind',
      'name',
      'energy',
      'token',
      'payment',
      'image',
      'maxUses',
      'learn',
      'copyable',
      'stable',
      'meme',
      'note',
      'effects',
      'resourceCosts',
    ],
    ['gold'],
  );
  const fx = exactRecord(r.effects, ['base', 'I', 'II']);
  return {
    id: label(r.id),
    owner: enumValue(r.owner, [...ROLES, '全局公共'], 'owner'),
    kind: enumValue(r.kind, SKILL_KINDS, 'skill kind'),
    name: label(r.name),
    energy: amount(r.energy),
    token: amount(r.token),
    payment: enumValue(r.payment, ['mix', 'api', 'sub'], 'payment'),
    image: amount(r.image),
    maxUses: nullable(r.maxUses, amount),
    learn: amount(r.learn),
    copyable: booleanValue(r.copyable, 'copyable'),
    stable: booleanValue(r.stable, 'stable'),
    meme: prose(r.meme),
    note: prose(r.note),
    effects: { base: effect(fx.base), I: effect(fx.I), II: effect(fx.II) },
    resourceCosts: dictionary(r.resourceCosts, amount, 16),
    gold: r.gold === undefined ? undefined : amount(r.gold),
  };
}
function buff(value: unknown): Buff {
  const r = exactRecord(
    value,
    ['id', 'name', 'kind', 'p', 'q', 'n', 'cap', 'trigger', 'expiry', 'overwrite', 'target'],
    [
      'category',
      'source',
      'description',
      'meme',
      'stackGroup',
      'reapply',
      'cacheScope',
      'refreshOnCombo',
    ],
  );
  return {
    id: label(r.id),
    name: label(r.name),
    kind: label(r.kind),
    p: amount(r.p),
    q: amount(r.q),
    n: amount(r.n),
    cap: amount(r.cap),
    trigger: prose(r.trigger),
    expiry: prose(r.expiry),
    overwrite: prose(r.overwrite),
    target: prose(r.target),
    category: optionalText(r, 'category'),
    source: optionalText(r, 'source'),
    description: optionalText(r, 'description'),
    meme: optionalText(r, 'meme'),
    stackGroup: optionalText(r, 'stackGroup'),
    reapply: optionalText(r, 'reapply'),
    cacheScope: optionalText(r, 'cacheScope'),
    refreshOnCombo:
      r.refreshOnCombo === undefined ? undefined : booleanValue(r.refreshOnCombo, 'refresh combo'),
  };
}
const byID = (v: { id: string }) => v.id;
function modeCatalog(value: unknown, mode: 'quick' | 'standard'): ModeCatalog {
  const outer = exactRecord(value, [
    'rules_version',
    'design_version',
    'schema_version',
    'content_hash',
    'config',
  ]);
  safeInteger(outer.rules_version, 1, 1, 'rules');
  enumValue(outer.design_version, ['0.17.0'], 'design');
  safeInteger(outer.schema_version, 15, 15, 'schema');
  const r = exactRecord(outer.config, [
    'schemaVersion',
    'mode',
    'name',
    'parameters',
    'paramMeta',
    'rules',
    'roles',
    'skills',
    'buffs',
    'loadouts',
    'resources',
    'harnesses',
    'passives',
  ]);
  enumValue(r.mode, [mode], 'mode');
  safeInteger(r.schemaVersion, 15, 15, 'config schema');
  const rules = exactRecord(r.rules, [
    'cacheWindow',
    'uniqueSamples',
    'strictSamples',
    'imageShortage',
  ]);
  enumValue(rules.cacheWindow, ['round'], 'cache window');
  enumValue(rules.imageShortage, ['illegal'], 'image shortage');
  if (
    !booleanValue(rules.uniqueSamples, 'samples') ||
    !booleanValue(rules.strictSamples, 'samples')
  )
    invalidResponse('sample policy');
  const parameters = dictionary(r.parameters, amount);
  const parameterKeys = [
    'API_PACK',
    'API_PRICE',
    'API_START',
    'BURST_CAP',
    'BURST_PERIOD',
    'CHARGE_PACK',
    'CHARGE_PRICE',
    'CLEANSE_PRICE',
    'DS_API_PACK',
    'DS_API_START',
    'ENERGY_CAP',
    'ENERGY_FLOOR',
    'ENERGY_START',
    'INITIAL_GOLD',
    'INSERT_CAP',
    'MAX_ROUNDS',
    'POWER_GENERATION',
    'PREP_MAX',
    'REGULATOR_PRICE',
    'REGULATOR_SAVE',
    'SUB_BURST_UPGRADE',
    'SUB_PERIOD',
    'SUB_PRICE',
    'SUB_START',
    'SUB_TOTAL_UPGRADE',
    'TARGET_LIKES',
    'TOKEN_FLOOR',
    'TURN_SECONDS',
  ];
  exactRecord(parameters, parameterKeys);
  if (
    !parameters.MAX_ROUNDS ||
    !parameters.TARGET_LIKES ||
    !parameters.ENERGY_CAP ||
    parameters.TURN_SECONDS !== 20 ||
    parameters.PREP_MAX !== 2 ||
    parameters.INSERT_CAP !== 1
  )
    invalidResponse('rule parameters');
  const roles = unique(
    r.roles,
    5,
    (v) => {
      const q = exactRecord(v, [
        'id',
        'name',
        'focus',
        'difficulty',
        'note',
        'weakness',
        'overrides',
        'resources',
      ]);
      return {
        id: roleID(q.id),
        name: label(q.name),
        focus: prose(q.focus),
        difficulty: label(q.difficulty),
        note: prose(q.note),
        weakness: prose(q.weakness),
        overrides: dictionary(q.overrides, amount),
        resources: dictionary(q.resources, (v) => {
          const a = exactRecord(v, ['initial', 'cap']);
          return { initial: amount(a.initial), cap: amount(a.cap) };
        }),
      };
    },
    byID,
  );
  const skills = unique(r.skills, 48, skill, byID),
    buffs = unique(r.buffs, 46, buff, byID);
  const harnesses = unique(
    r.harnesses,
    8,
    (v) => {
      const q = exactRecord(v, ['id', 'name', 'activeSlots', 'passives', 'description', 'meme']);
      return {
        id: label(q.id),
        name: label(q.name),
        activeSlots: safeInteger(q.activeSlots, 0, 2, 'harness slots'),
        passives: list(q.passives, 2, label),
        description: prose(q.description),
        meme: prose(q.meme),
      };
    },
    byID,
  );
  const passives = unique(
    r.passives,
    8,
    (v) => {
      const q = exactRecord(v, ['id', 'name', 'kind', 'p', 'q', 'description', 'meme'], ['buffId']);
      return {
        id: label(q.id),
        name: label(q.name),
        kind: label(q.kind),
        p: amount(q.p),
        q: amount(q.q),
        description: prose(q.description),
        meme: prose(q.meme),
        buffId: optionalText(q, 'buffId'),
      };
    },
    byID,
  );
  const loadouts = unique(
    r.loadouts,
    32,
    (v) => {
      const q = exactRecord(v, ['id', 'role', 'name', 'skills', 'note']);
      return {
        id: label(q.id),
        role: label(q.role),
        name: label(q.name),
        skills: unique(q.skills, 6, label, (v) => v),
        note: prose(q.note),
      };
    },
    byID,
  );
  const resources = unique(
    r.resources,
    16,
    (v) => {
      const q = exactRecord(
        v,
        ['id', 'name', 'unit', 'description', 'meme'],
        ['subscription', 'upgrade'],
      );
      return {
        id: label(q.id),
        name: label(q.name),
        unit: label(q.unit),
        description: prose(q.description),
        meme: prose(q.meme),
        subscription:
          q.subscription === undefined ? undefined : booleanValue(q.subscription, 'subscription'),
        upgrade: q.upgrade === undefined ? undefined : amount(q.upgrade),
      };
    },
    byID,
  );
  const paramMeta = unique(
    r.paramMeta,
    128,
    (v) => {
      const q = exactRecord(v, ['id', 'name', 'unit', 'note']);
      return { id: label(q.id), name: label(q.name), unit: prose(q.unit), note: prose(q.note) };
    },
    byID,
  );
  if (
    roles.length !== 5 ||
    skills.length !== 48 ||
    buffs.length !== 46 ||
    harnesses.length !== 8 ||
    passives.length !== 8 ||
    resources.length !== 1
  )
    invalidResponse('catalog completeness');
  for (const sk of skills)
    for (const fx of [sk.effects.base, sk.effects.I, sk.effects.II])
      for (const id of [fx.buffId, fx.extraBuffId])
        if (id && !buffs.some((b) => b.id === id)) invalidResponse('buff reference');
  for (const h of harnesses)
    if (h.passives.some((id) => !passives.some((p) => p.id === id)))
      invalidResponse('passive reference');
  for (const p of loadouts)
    if (
      p.skills.length < 1 ||
      p.skills.length > 4 ||
      p.skills.some(
        (id) => !skills.some((s) => s.id === id && (s.owner === p.role || s.owner === '全局公共')),
      )
    )
      invalidResponse('preset reference');
  for (const role of roles)
    if (
      !loadouts.some((p) => p.role === role.id) ||
      skills.filter((s) => s.owner === role.id).length !== 8
    )
      invalidResponse('role completeness');
  if (
    skills.filter((s) => s.owner === '全局公共').length !== 8 ||
    paramMeta.length !== parameterKeys.length ||
    paramMeta.some((p) => !(p.id in parameters))
  )
    invalidResponse('catalog completeness');
  return {
    contentHash: hashValue(outer.content_hash),
    mode,
    name: label(r.name),
    parameters,
    paramMeta,
    roles,
    skills,
    buffs,
    harnesses,
    passives,
    loadouts,
    resources,
  };
}
export function likesCatalog(value: unknown): LikesCatalog {
  const r = exactRecord(value, [
    'rules_version',
    'design_version',
    'schema_version',
    'content_hash',
    'modes',
  ]);
  safeInteger(r.rules_version, 1, 1, 'catalog rules');
  enumValue(r.design_version, ['0.17.0'], 'catalog design');
  safeInteger(r.schema_version, 15, 15, 'catalog schema');
  const modes = exactRecord(r.modes, ['quick', 'standard']);
  return {
    contentHash: hashValue(r.content_hash),
    modes: {
      quick: modeCatalog(modes.quick, 'quick'),
      standard: modeCatalog(modes.standard, 'standard'),
    },
  };
}
